src/tcp/DataBlocks.jl

```diff
@@ -96,19 +102,91 @@ end
 
 function chread(sock::ClickHouseSock, ::Type{Block})::Block
     temp_table = chread(sock, String)
-    block_info = chread(sock, BlockInfo)
-    num_columns = chread(sock, VarUInt)
-    num_rows = chread(sock, VarUInt)
-    columns = [read_col(sock, num_rows) for _ ∈ 1:UInt64(num_columns)]
-    Block(temp_table, block_info, num_columns, num_rows, columns)
+    main_io = sock.io
+    try
+        if compression_enabled(sock.settings)
+            hash = chread(sock, UInt128)
+            method = chread(sock, Compression)
+            raw_len = chread(sock, UInt32)
+            data_len = chread(sock, UInt32)
+            compressed_len = VarUInt(raw_len - HEADER_SIZE_W_COMPRESSION)
+            compressed = chread(sock, Vector{UInt8}, compressed_len)
+
+            # check packet checksum
+            packet = [
+                UInt8(method);
+                reinterpret(UInt8, [raw_len]) |> Vector{UInt8};
+                reinterpret(UInt8, [data_len]) |> Vector{UInt8};
+                compressed
+            ]
```

Comment:
Exceptionally slow & allocation heavy. Need resolve

Reply:
Simplest approach that comes to mind is to just alloc a vector of size `compressed_len + HEADER_SIZE_W_COMPRESSION` with uninitialized memory up front, write the required header prefix to the new memory, get rid of

https://github.com/JuliaDatabases/ClickHouse.jl/blob/03def1ec7dac9a1567cfc4715159ba2edffde7d9/src/tcp/DataBlocks.jl#L113

and instead simply do a manual `read!` call, passing it `@view new_arr[HEADER_SIZE_W_COMPRESSION:]`. This view can then also be passed to `decompress` and we should end up with 0 additional copies without having to alter the CityHash implementation expose a block/iterator-based API.